testsuite/tests/typing-poly/poly.ml

```diff
@@ -1988,3 +1987,53 @@ Error: This pattern matches values of type [? `Y ]
        The second variant type is bound to the universal type variable 'b,
        it may not allow the tag(s) `Y
 |}]
+
+let fail_example_corrected: 'a . 'a -> [< `X of 'a | `Y > `X ] -> 'a = fun x y ->
+  match y with
+  | `Y -> x
+  | `X x -> x
+[%%expect {|
+val fail_example_corrected : 'a -> [< `X of 'a | `Y > `X ] -> 'a = <fun>
+|}]
```

Comment:
Note: on trunk and your branch, I think that you can actually write the more general argument ``[< `X of 'a | `Y ]`` (as I did in my review comment), as it does not introduce a weird intersection as it did on 5.0 and earlier.